File: frontend/src/types/chatContract.ts
/** Kontrakt API czatu — zgodny z schemas/chat_contract.py */

import type { HearingRound } from '../store/useTrialRoomStore';
import type {
  ClaimScore,
  ExpertAnalysis,
  InvestigationSummary,
  SourceReference,
  StepDiagnostic,
  ChatMessage,
} from './chat';

export type ChatMode = 'single' | 'moa' | 'consensus';

export type ResponseMode = 'citizen' | 'strategic' | 'draft';

export type ProcessSide = 'defense' | 'prosecution';

export interface PromptOverrides {
  architect_prompt?: string;
  system_role_prompt?: string;
  judge_system_prompt?: string;
  task_prompt?: string;
  role_catalog?: Record<string, string>;
  expert_role_prompts?: Record<string, string>;
}

export interface MoaOptions {
  selected_models: string[];
  aggregator_model: string;
  expert_roles_map: Record<string, string>;
}

export interface ChatAttachment {
  id?: string;
  name?: string;
  type?: string;
  mime_type?: string;
  content_type?: string;
  url?: string;
  text?: string;
  size?: number;
  content?: unknown;
}

export interface ChatHistoryMessage {
  role: string;
  content: string | unknown[] | Record<string, unknown>;
}

export interface ChatPayloadV2 {
  message: string;
  chat_mode: ChatMode;
  response_mode: ResponseMode;
  side: ProcessSide;
  active_system_role_id?: string;
  current_task?: string;
  prompt_overrides: PromptOverrides;
  moa_options?: MoaOptions;
  model?: string;
  sessionId?: string;
  attachments?: ChatAttachment[];
  document_text?: string;
  history?: ChatHistoryMessage[];
  act_terms?: string[];
  use_saos?: boolean;
  use_eli?: boolean;
  use_rag_legal?: boolean;
  use_rag_user?: boolean;
  use_lexminde_mcp?: boolean;
  model_latencies?: Record<string, number>;
  stream?: boolean;
  active_prompt_preset_id?: string;

  // Transitional compatibility fields. Keep until backend legacy adapter is removed.
  selected_models?: string[];
  aggregator_model?: string;
  expert_roles?: Record<string, string>;
  architect_prompt?: string;
  system_role_prompt?: string;
  judge_system_prompt?: string;
  task_prompt?: string;
  role_catalog?: Record<string, string>;
  expert_role_prompts?: Record<string, string>;
}

export interface ChatMutationExtras {
  attachments?: ChatAttachment[];
  sessionId?: string;
  document_text?: string;
  history?: ChatHistoryMessage[];
  use_saos?: boolean;
  use_eli?: boolean;
  use_rag_legal?: boolean;
  use_rag_user?: boolean;
  use_lexminde_mcp?: boolean;
  act_terms?: string[];
}

export interface StreamMetadataEvent {
  type: 'metadata';
  id?: string;
  sessionId?: string;
  message?: string;
  expert_analyses?: ExpertAnalysis[];
  urgency_alerts?: unknown[];
}

export interface StreamChunkEvent {
  type: 'chunk';
  text: string;
}

export interface StreamFinalMetadataEvent {
  type: 'final_metadata';
  id?: string;
  sessionId?: string;
  final_answer?: string;
  sources?: string[];
  expert_analyses?: ExpertAnalysis[];
  eli_explanation?: string;
  diagnostics?: StepDiagnostic[];
  pipeline_latency_ms?: number;
  urgency_alerts?: unknown[];
  timeline?: ChatMessage['timeline'];
  gaps?: string[];
  inconsistencies?: string[];
  coi_conflicts?: string[];
  p_sukces?: number;
  confidence_score?: number;
  hitl_escalated?: boolean;
  synthesis_blocked?: boolean;
  hallucinated_cites?: string[];
  claim_scores?: ClaimScore[];
  investigation_summary?: InvestigationSummary;
  cited_sources?: SourceReference[];
  hearing_rounds?: HearingRound[];
}

export interface StreamErrorEvent {
  type: 'error';
  text?: string;
}

export interface TrialRoundEvent {
  type: 'trial_round';
  side?: string;
  round?: number | string;
  text?: string;
  model?: string;
}

export type ChatStreamEvent =
  | StreamMetadataEvent
  | StreamChunkEvent
  | StreamFinalMetadataEvent
  | StreamErrorEvent
  | TrialRoundEvent;

File: schemas/chat_contract.py
"""Żelazny kontrakt żądania czatu (API v2)."""
from __future__ import annotations

from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import AliasChoices, BaseModel, Field, field_validator


class ChatMode(str, Enum):
    SINGLE = "single"
    MOA = "moa"
    CONSENSUS = "consensus"

    @classmethod
    def normalize(cls, value: Optional[str]) -> "ChatMode":
        raw = (value or "single").strip().lower()
        if raw in ("moa", "consensus"):
            return cls.MOA
        return cls.SINGLE

    @property
    def is_moa(self) -> bool:
        return self in (ChatMode.MOA, ChatMode.CONSENSUS)


class ResponseMode(str, Enum):
    CITIZEN = "citizen"
    STRATEGIC = "strategic"
    DRAFT = "draft"

    @classmethod
    def normalize(cls, value: Optional[str]) -> "ResponseMode":
        raw = (value or "strategic").strip().lower()
        if raw in ("default", "advisor"):
            return cls.STRATEGIC
        if raw == "citizen":
            return cls.CITIZEN
        if raw == "draft":
            return cls.DRAFT
        return cls.STRATEGIC


class ProcessSide(str, Enum):
    DEFENSE = "defense"
    PROSECUTION = "prosecution"

    @classmethod
    def normalize(cls, value: Optional[str]) -> "ProcessSide":
        raw = (value or "defense").strip().lower()
        if raw == "prosecution":
            return cls.PROSECUTION
        return cls.DEFENSE


class LegalSourceType(str, Enum):
    CONSTITUTION = "constitution"
    STATUTE = "statute"
    REGULATION = "regulation"
    CASE_LAW = "case_law"
    USER_DOC = "user_doc"


class PromptOverrides(BaseModel):
    architect_prompt: Optional[str] = None
    system_role_prompt: Optional[str] = None
    judge_system_prompt: Optional[str] = None
    task_prompt: Optional[str] = None
    role_catalog: Optional[Dict[str, str]] = None
    expert_role_prompts: Optional[Dict[str, str]] = None

    model_config = {"extra": "ignore"}


class MoaOptions(BaseModel):
    selected_models: List[str] = Field(default_factory=list)
    aggregator_model: str = ""
    expert_roles_map: Dict[str, str] = Field(default_factory=dict)

    model_config = {"extra": "ignore"}


class ChatAttachment(BaseModel):
    name: Optional[str] = None
    type: Optional[str] = None
    mime_type: Optional[str] = None
    content_type: Optional[str] = None
    url: Optional[str] = None
    text: Optional[str] = None
    size: Optional[int] = None
    content: Optional[Any] = None

    model_config = {"extra": "allow"}


class ChatHistoryMessage(BaseModel):
    role: str
    content: Any

    model_config = {"extra": "allow"}


class ChatPayloadV2(BaseModel):
    message: str = ""
    chat_mode: ChatMode = ChatMode.SINGLE
    response_mode: ResponseMode = ResponseMode.STRATEGIC
    side: ProcessSide = ProcessSide.DEFENSE
    active_system_role_id: Optional[str] = None
    current_task: Optional[str] = None
    prompt_overrides: PromptOverrides = Field(default_factory=PromptOverrides)
    moa_options: Optional[MoaOptions] = None

    model: Optional[str] = None
    session_id: Optional[str] = Field(
        default=None,
        validation_alias=AliasChoices("session_id", "sessionId", "sid"),
    )
    attachments: List[ChatAttachment] = Field(default_factory=list)
    document_text: Optional[str] = None
    history: List[ChatHistoryMessage] = Field(default_factory=list)
    act_terms: Optional[List[str]] = None

    use_saos: bool = True
    use_eli: bool = True
    use_rag_legal: bool = True
    use_rag_user: Optional[bool] = None
    model_latencies: Optional[Dict[str, float]] = None

    model_config = {"extra": "ignore", "populate_by_name": True}

    @field_validator("message")
    @classmethod
    def message_not_empty_for_chat(cls, v: str) -> str:
        return v or ""
